crates/wascc-host/src/actors/actor_host.rs

```diff
@@ -10,35 +10,49 @@ use futures::executor::block_on;
 use wapc::{WapcHost, WasiParams};
 use wascap::prelude::{Claims, KeyPair};
 
+#[derive(Default)]
 pub(crate) struct ActorHost {
+    state: Option<State>,
+}
+
+struct State {
     guest_module: WapcHost,
     claims: Claims<wascap::jwt::Actor>,
     mw_chain: Vec<Box<dyn Middleware>>,
     image_ref: Option<String>,
 }
 
-impl ActorHost {
-    pub fn new(
-        actor_bytes: Vec<u8>,
-        wasi: Option<WasiParams>,
-        mw_chain: Vec<Box<dyn Middleware>>,
-        signing_seed: String,
-        image_ref: Option<String>,
-    ) -> ActorHost {
-        let buf = actor_bytes.clone();
-        let actor = WasccActor::from_slice(&buf).unwrap();
+#[derive(Message)]
+#[rtype(result = "Result<()>")]
+pub(crate) struct Initialize {
+    pub actor_bytes: Vec<u8>,
+    pub wasi: Option<WasiParams>,
+    pub mw_chain: Vec<Box<dyn Middleware>>,
+    pub signing_seed: String,
+    pub image_ref: Option<String>,
+}
+
+impl Handler<Initialize> for ActorHost {
+    type Result = Result<()>;
+
+    fn handle(&mut self, msg: Initialize, ctx: &mut Self::Context) -> Self::Result {
+        let buf = msg.actor_bytes.clone();
+        let actor = WasccActor::from_slice(&buf)?;
 
         #[cfg(feature = "wasmtime")]
         let engine = wasmtime_provider::WasmtimeEngineProvider::new(&buf, wasi);
         #[cfg(feature = "wasm3")]
         let engine = wasm3_provider::Wasm3EngineProvider::new(&buf);
 
         let c = actor.token.claims.clone();
+        let c2 = c.clone();
+        let c3 = c.clone(); // TODO: I can't believe I have to do this to make the [censored] borrow checker happy
+        let seed = msg.signing_seed.to_string();
 
         let guest = WapcHost::new(Box::new(engine), move |_id, bd, ns, op, payload| {
             crate::dispatch::wapc_host_callback(
-                KeyPair::from_seed(&signing_seed).unwrap(),
-                actor.token.claims.clone(),
+                KeyPair::from_seed(&seed).unwrap(),
+                c2.clone(),
```

Comment:
This needs to be cloned again too? Man this borrow checker 😢

Reply:
It's really disgusting how many times I have to build a "preamble" full of clones all so that I can move detached versions of self-owned fields into a closure.

Reply:
It's harder in some places with actix because `block_on(async move { ... }))` is actually a _double closure_, since the input to `block_on` is a closure, and then inside the async block is also a closure.

Reply:
Wild. Good to know